File: pset6/credit/credit.py
from cs50 import get_int
import re


def main():
    # Get credit card number from user input
    card_number = get_int("Number: ")

     # Check if the card number is valid using the Luhn's algorithm
    if luhns(card_number):
        # Print the type of card if valid
        print(kind_of_card(card_number))
    else:
        # Print INVALID if Luhn's check fails
        print("INVALID")

# Luhn's algorithm to validate credit card number
def luhns(n):

    sum = 0
    counter = 0
    
    # Loop through digits in reverse (right to left)
    while n > 0:
        last_digit = n % 10  # Get last digit
        n //= 10  # Remove last digit

        # Double every second digit starting from the right
        if counter % 2 == 1:
            doubled = last_digit * 2
            if doubled > 9:
                sum += doubled - 9
            else:
                sum += doubled
        else:
            sum += last_digit

        counter += 1

    # Valid if total sum modulo 10 is 0
    return sum % 10 == 0


def kind_of_card(n):
    #  Use regex patterns to determine credit card type

    # VISA cards start with 4 and can be 13 or 16 digits
    visa_pattern = r"^4\d{12}(\d{3})?(\d{3})?$"
    mastercard_pattern = r"^5[1-5]\d{14}$"  # MASTERCARD starts with 51-55 and is 16 digits
    amex_pattern = r"^3[47]\d{13}$"  # AMEX cards start with 34 or 37 and are 15 digits

    # Convert card number to string for regex match
    n = str(n)

    # Match card number to known patterns and return the type
    if re.match(visa_pattern, n):
        return "VISA"
    elif re.match(mastercard_pattern, n):
        return "MASTERCARD"
    elif re.match(amex_pattern, n):
        return "AMEX"
    else:
        return "INVALID"

# Run the program if this file is executed directly
if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int luhns(long number);
void kind_of_card(long number);

int main(void)
{
    long card_number = get_long("Number: ");    // Prompt for card number
    int sum = luhns(card_number);    // Get sum for Luhn's algorithm
    sum %= 10;    // Get last digit    
    if (sum == 0)    // If valid
    {
        kind_of_card(card_number);    // Identify card type
    }
    else // If invalid
    {
        printf("INVALID\n");
    }
}

// Luhn's algorithm
int luhns(long number)
{
    int counter = 0, sum = 0, sum2 = 0;
    while (number > 0)
    {
        int last_digit = number % 10;    // Get last digit
        number /= 10;    // Remove last digit
        if (counter == 1)    // For every second digit (from second to last)
        {
            int doubling;
            doubling = last_digit * 2;
            if (doubling > 9)
            {
                sum = sum + 1 + (doubling % 10);    // Add sum of digits of doubled value
            }
            else
            {
                sum = sum + doubling;    // Add double value
            }
            counter = 0;
        }
        else    // For other digits
        {
            sum2 = sum2 + last_digit;    // Add digit to sum2
            counter++;
        }
    }
    sum += sum2;    // Final sum
    return sum;
}

// Determine type of card based on first digits and length
void kind_of_card(long number)
{
    int counter = 0;
    long num = number;

    // Count number of digits
    while (num > 0)
    {
        int last_digit = num % 10;
        num /= 10;
        counter++;
    }
    
    int first_digit;
    num = number;

    // Remove first digit
    while (num >= 10)
    {
        num /= 10;
    }
    first_digit = num;
    
    num = number;
    int first_two_digits;

    // Remove first two digits
    while (num >= 100)
    {
        num /= 10;
    }

    first_two_digits = num;

    // Check card type
    if ((first_digit == 4 && counter == 13) || (first_digit == 4 && counter == 16))
    {
        printf("VISA\n");
    }
    else if ((first_two_digits == 34 && counter == 15) || (first_two_digits == 37 && counter == 15))
    {
        printf("AMEX\n");
    }
    else if (first_two_digits >= 51 && first_two_digits <= 55 && counter == 16)
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}
